core/src/main/scala/scala/pickling/generator/algorithms.scala

```diff
@@ -0,0 +1,85 @@
+package scala.pickling
+package generator
+
+import scala.reflect.api.Universe
+
+// TODO - These need logging messages.
+
+/** An interface so we can pass logging to these algorithms at runtime/during testing. */
+trait AlgorithmLogger {
+  def warn(msg: String): Unit
+  def debug(msg: String): Unit
+  def error(msg: String): Unit
+  def abort(msg: String): Nothing
+}
+
+/** Represents the result of an algorithm call. */
+sealed trait AlgorithmResult {
+  def join(other: => AlgorithmResult): AlgorithmResult
```

Comment:
Given that your success case returns `this`, isn't this more like `Option`'s `orElse`?
In FP (by which I mean Haskell) `join` means `flatten`.